source/extensions/tracers/xray/util.cc

```diff
@@ -0,0 +1,51 @@
+#include "extensions/tracers/xray/util.h"
+
+#include "absl/strings/ascii.h"
+
+namespace Envoy {
+namespace Extensions {
+namespace Tracers {
+namespace XRay {
+
+// TODO(magmarco): fuzz test this function
```

Comment:
I am wondering whether we can converge with other matchers in Envoy, e.g. RE2 regexes. New parsers and languages are discouraged where we can make this work with some existing setup.

Reply:
This is a lot simpler _and different_ from a regex though.
For example a* in a regex matches zero or more 'a' while in wildcard it matches the letter 'a' followed by anything (or nothing).

That being said, it might be a good idea to move this to Envoy's common matchers as a utility function. Potentially beefing it up a bit with an optional cases-sensitivity flag

Reply:
Also it could be made faster a bit by compiling the pattern.

Reply:
I think we want to really make this a lot more robust. With my security hat on, I really dislike seeing custom parsers being introduced to the Envoy code base unless totally necessary. If we're going to add this, I suggest we have this made a standard matcher and implement it with an extensive set of tests and fuzzers. Is there any standard library we could use instead?

Reply:
>If we're going to add this, I suggest we have this made a standard matcher and implement it with an extensive set of tests and fuzzers.

Fuzzing this method is a reasonable ask, that's why I've added the TODO. If we want to make this function part of
envoy's utility functions, that's great but hopefully that would not block this PR from getting merged. After all,
this PR is a **skeleton** for x-ray support. I still have the bulk of the code to be submitted for review. I will own
the x-ray plugin moving forward (even though I work at the AppMesh team and not X-Ray) so I'll be happy to refactor it when we have a function in Envoy that does the same thing.

>Is there any standard library we could use instead?

Not sure what standard here refers to, but definitely nothing in the C++ standard, if there's a similar function in abseil, I'll be happy to use it instead.


Please consider not blocking this PR because of this function.

Reply:
@jmarantz I've added fuzz tests btw.

Reply:
Thanks, I feel better about this with some fuzzing. Can you articulate for purpose of review why we can't just transform the glob input into a RE2 regex? It seems that any any glob can be turned into a regex (but not the other way around). That way, the code is only a few lines (some `std::regex_replace` on the glob, build a RE2 regex and then match). There is a performance cost, but I think it's not worth optimizing for without data showing this is a major issue.

Reply:
Basically, users of X-Ray can specify rules to decide which requests get traced. The way they specify the rules can include wildcards. That's not something we can change.

See [Sampling Rules in X-Ray](https://docs.aws.amazon.com/en_pv/xray/latest/devguide/xray-console-sampling.html#xray-console-sampling-options)
> String values can use wildcards to match a single character (?) or zero-or-more characters (*).

I think it's a good idea like you said to move this function to Envoy's common utilities. But I hope that doesn't block merging this PR. We need to keep moving with the rest of XRay's implementation.

Reply:
I'm not asking to change the syntax. I'm suggesting (or asking why we can't) translate from globs to RE2 internally. It seems a simple transform that gets rid of this hard to reason about code.

Reply:
I'm saying that the pattern is a user input. If you're suggesting we take the user input and convert it to regex and then do matching, then that's going to be significantly slower. This matching is going to happen on _every single request_ to determine whether to sample the request or not.

That being said, what's the reason for pushing back so hard on this ~10 line function?
It's tested and fuzzed and I don't think it's hard to reason about honestly.
Let's discuss this offline over slack.

Reply:
At config ingestion command you can (& IMO should) compile the wildcard into something that can be matched more quickly, rather than doing that every request. Maybe by RE, maybe by heavily scrutinized custom code.

I'm not sure you want untrusted input going into std::regex; It can OOM (https://github.com/envoyproxy/envoy/issues/7728).  This is config not traffic so maybe it's OK?

Alternatively you could use RE2 for transforming the wildcard-to-regex at config time, or maybe even some very simple absl string substitution calls to do the transform from wildcard to RE2?

In any case I'd recommend running a microbenchmark, maybe patterned after https://github.com/envoyproxy/envoy/blob/master/test/common/common/utility_speed_test.cc

There's also a benchmark I have from the past, for a group of wildcards, but the benchmarking API has evolved since then: https://github.com/apache/incubator-pagespeed-mod/blob/master/pagespeed/kernel/base/fast_wildcard_group_speed_test.cc

Reply:
OK - I've talked with @htuch offline and I'm pasting the rationale here for posterity.

AWS X-Ray has two sources of sampling rules (wildcard patterns):
1. The manifest file (static config)
1. The X-Ray Service - periodically polled for new rules. (dynamic config)

The other part is that X-Ray will inspect every single request and try to match it against the set of rules it has to decide whether or not a given request should be sampled. That means this wildcard matching routine is on the hot path. I've spent a great deal of time to make this function optimal and not allocate. Converting the patterns to regex means we need to do allocations for string replacements, compile the regex and then match. That will be significantly slower than the current wildcard matcher.

Moreover, I've added a comprehensive set of test cases to this routine and ASAN should give us a pretty good guarantee that it's not doing out of bounds access or any funny stuff of that sort. I've also added fuzz tests, and I've also had this function reviewed by you guys and by members of my team.
So, from a security perspective, I think we're looking good.

I'm going to add comments to the function to document the reasoning above.
@htuch will take the time and review this function more closely.

Reply:
Sounds good. I still don't understand why compiling the expression is in the hot-path. It just happens from manifest once, plus periodic x-ray polling whenever the input actually changes.

Reply:
If you look at the [Sampling Rules in X-Ray](https://docs.aws.amazon.com/en_pv/xray/latest/devguide/xray-console-sampling.html#xray-console-sampling-options) you will see that there are multiple parts to each rule that can have a pattern. That means I would have to a) create a regex by replacing the input to the corresponding regex, b)compile the regex, and c)store a regex for each corresponding part in every rule. That's an overkill to avoid a function that is well tested and does zero allocations.

More importantly, think about all the corner cases of the user input pattern that might happen to contain legitimate regex syntax(e.g. '.'), which can be an attack vector if we get wrong. 
I'm pretty sure it's a lot easier to verify the correctness of this wildcard matcher versus verifying the correctness of a regex linter/escape function.

Reply:
I agree your algorithm looks acceptable.

What I would prefer at some point though is having a Wildcard class you could re-populate whenever the input pattern changes. Again, not for this PR. But if we wind up pulling something into the general Envoy matcher class then I'd want it look like that so we can optimize it more by examining the pattern(s) before requests start flowing.

Reply:
sounds good.